lib/ukdebug/README.md

```diff
@@ -0,0 +1,201 @@
+command with `-serial stdio`, you should see some output that ends with
+something like the below (assuming you have debug messages enabled):
+
+```plain
+[    0.104746] Info: [libukdebug] <gdbstub.c @ 1237> Waiting for debugger connection on con0...
+[    0.104944] Info: [libukconsole] <console.c @  176> Registered con1: GDB virtual console, flags: --
+$S05#b8
+```
+
+The last line is the first packet that the GDB stub sends to the GDB host.
+Switch back to `-serial tcp::1235,server,nowait` and you'll stop seeing any
+output. Using `-serial stdio` can be useful when you're unsure if your
+configured kernel even boots.
+
+Note that the `-serial` flag is position-sensitive on x86. COM1 is connected
+to whatever the first occurrence (from left to right) of the `-serial` flag
+specifies. COM2 is configured based on the second occurrence of `-serial`, and
+so forth. E.g., as an alternative to the above, you could configure COM1 and
+COM2 in kconfig and then connect the GDB stub to COM2 and keep kernel output on
+COM1 (`debug.gdbcon` is `1` this time):
+
+```bash
+qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
+                   -cpu max \
+                   -display none \
+                   -serial stdio \
+                   -serial tcp::1235,server,nowait \
+                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=1 --"
+```
+
+But let's stick to the initial example and assume that only COM1 is configured
+and that `debug.gdbcon=0`. Start the VM and in another terminal run GDB like
+this:
+
+```bash
+gdb build/app-helloworld_qemu-x86_64.dbg
+```
+
+There are two things to note here. First, don't forget the `.dbg` extension.
+GDB won't find any symbols without it. Second, make sure you're using the right
+GDB binary. E.g., if you're developing on an x86 system and the kernel is
+compiled for Arm, you need to use GDB for Arm, not the stock GDB of your system.
+
+Run the command `target remote localhost:1235` in the GDB REPL to connect to
+the GDB stub on TCP socket `localhost:1235`. GDB should manage to connect to
+the VM and you should see where in the kernel execution was stopped. E.g., it
+might look like this:
+
+```plain
+(gdb) tar remote :1235
+Remote debugging using :1235
+0x0000000000120e02 in gdb_entry (ctx=<optimized out>) at /home/thasso/Work/oss-unikraft/lib/ukdebug/gdbstub.c:1247
+1247    }
+```
+
+You are now connected to the GDB stub that runs inside Unikraft. All the
+usual GDB commands like `break`, `backtrace`, or `continue` should work. Since
+we only configured a single console and didn't set `debug.gdbmode` on the
+command line, the GDB should automatically share the backing console for kernel
+output. If you `break main` and `continue`, you should see all kernel output
+printed inline with the other output from GDB.
+
+### Debugging the GDB stub
+
+This is guidance for anyone working on the GDB stub who needs to debug it.
```

Comment:
> This is guidance for anyone working on the GDB stub who needs to debug it.

"This section provides information for developers who need to debug Unikraft's GDB stub."